text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,174 @@
+* Start Date: 2018-08-01
+* RFC PR: https://github.com/fusionjs/rfcs/pull/8
+
+# Summary
+
+This RFC describes a process for standardizing on an API for serializing and deserializing state between server and browser plugins.
+
+# Basic example
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    const {serialize, deserialize} = SharedState('some-id');
+    return (ctx, next) => {
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize(ctx, {
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+
+// in tests
+import {SharedStateToken, MockSharedState} from 'fusion-core';
+app.register(SharedStateToken, MockSharedState({
+  'some-id': {
+    some: 'json'
+  },
+  'other-data': {}
+}));
+```
+
+# Motivation
+
+It is a very common pattern in fusion to have some shared state between the server and
+the browser encapsulated into a plugin. The current model for this has users referencing
+`ctx.template` directly on the server, and querying the DOM directly from the browser.
+
+This can make testing difficult, especially when running browser tests since there is not
+an easy way to mock the deserialization other than each plugin exposing their own deserialization
+token. 
+
```

Comment:
Nice idea! I think generally it's good if we move things that do side effects into stuff that is injectable (for testing). Hopefully this would make it easier for more plugins to be totally pure.

That said, for testing I think long-term hopefully we can make the puppeteer-style testing easier so not as much needs to be mocked.

Reply:
Yea, a separate issue from this is having puppeteer style tests with access to `app`, but still I think this is useful even when we have support for that.